src/Adapter/Order/CommandHandler/AddProductToOrderHandler.php

```diff
@@ -498,27 +440,163 @@ private function createNewInvoice(Order $order, Cart $cart, $isFreeShipping)
+        $orderTotals = OrderTotalNumbers::buildFromOrder($order);
+
+        // update totals amount of order
+        $order->total_products = (float) (string) $orderTotals->getTotalProducts()
+            ->plus($this->number($cart->getOrderTotal(true, $totalMethod)))
+        ;
+        $order->total_products_wt = (float) (string) $orderTotals->getTotalProductsWt()
+            ->plus($this->number($cart->getOrderTotal(true, Cart::ONLY_PRODUCTS)))
+        ;
+
+        $paidWithTaxes = $this->number($cart->getOrderTotal(true, $totalMethod));
+        $order->total_paid = (float) (string) $orderTotals->getTotalPaid()->plus($paidWithTaxes);
+        $order->total_paid_tax_incl = (float) (string) $orderTotals->getTotalPaidTaxIncl()->plus($paidWithTaxes);
+        $order->total_paid_tax_excl = (float) (string) $orderTotals->getTotalPaidTaxExcl()
+            ->plus($this->number($cart->getOrderTotal(false, $totalMethod)))
+        ;
+
+        // discount
+        $discountWithTaxes = $this->number($cart->getOrderTotal(true, Cart::ONLY_DISCOUNTS));
+        $order->total_discounts = (float) (string) $orderTotals->getTotalDiscounts()->plus($discountWithTaxes);
+        $order->total_discounts_tax_incl = (float) (string) $orderTotals->getTotalDiscountTaxIncl()->plus($discountWithTaxes);
+        $order->total_discounts_tax_excl = (float) (string) $orderTotals->getTotalDiscountTaxExcl()
+            ->plus($this->number($cart->getOrderTotal(false, Cart::ONLY_DISCOUNTS)))
+        ;
+
+        //wrapping
+        $wrappingWithTaxes = $this->number($cart->getOrderTotal(true, Cart::ONLY_WRAPPING));
+        $order->total_wrapping_tax_incl = (float) (string) $orderTotals->getTotalWrappingTaxIncl()->plus($wrappingWithTaxes);
+        $order->total_wrapping = (float) (string) $orderTotals->getTotalWrapping()->plus($wrappingWithTaxes);
+        $order->total_wrapping_tax_excl = (float) (string) $orderTotals->getTotalWrappingTaxExcl()
+            ->plus($this->number($cart->getOrderTotal(false, Cart::ONLY_WRAPPING)))
+        ;
+
+        //shipping
+        if ($isFreeShipping) {
+            $this->updateShippingManually($order, $cart, $orderTotals);
+        } else {
+            $order->refreshShippingCost();
+        }
+    }
+
+    /**
+     * couldn't use Order::refreshShippingCost() with free shipping because it recalculates whole cart shipping
+     *  and then total numbers doesn't fit together.
+     *  e.g initial shipping was $7 and discounts $0
+     *  we add product with free shipping (behind the scenes it adds a discount of shipping price ($7))
+     *  so for total prices to fit - shipping should be $14 and discounts $7
+     *  using  Order::refreshShippingCost() would end up shipping being $7
+     * @todo: If i do it manually, i don't update OrderCarrier shipping price.
+     *      Difference occurs between order shipping price and carrier shipping price
+     *      That might introduce problems. How To fix it?
```

Comment:
@matks I have a problem with this (explained in TODO). If its unclear ill explain in chat. Could you help?

Reply:
It would be easily solved if we apply shipping price to whole cart instead of one product. Or completely remove the option to create new invoice when adding product, or remove the free shipping option when adding product with new invoice. Idk all those features are very strange and doesn't comply with the system design

Reply:
this was discussed in standup and chat. Decided to leave is at it was in legacy and dive deeper in later versions.